crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/next_target.rs

```diff
@@ -0,0 +1,725 @@
+                            )
+                        })
+                })
+                .next();
+
+            // If we found a remote address add this agent to the node
+            // or create the node if it doesn't exist.
+            if let Some((agent, cert, url)) = info {
+                match remote_nodes.get_mut(&cert) {
+                    // Add the agent to the node.
+                    Some(node) => node.target.agents.push(agent),
+                    None => {
+                        // This is a new node.
+                        remote_nodes.insert(
+                            cert.clone(),
+                            Node {
+                                target: GossipTgt {
+                                    agents: vec![agent],
+                                    cert,
+                                },
+                                url,
+                            },
+                        );
+                    }
+                }
+            }
+        }
+
+        let remote_nodes = remote_nodes.into_iter().map(|(_, v)| v).collect();
+        let tuning_params = self.tuning_params.clone();
+        // We could clone the metrics store out of the lock here but I don't think
+        // the next_remote_node will be that slow so we can just choose the next node inline.
+        self.inner.share_mut(|i, _| {
+            let node = next_remote_node(remote_nodes, &i.metrics, tuning_params)
+                .map(|n| (n.target, n.url));
+            Ok(node)
+        })
+    }
+}
+
+/// Find the next remote node to sync with.
+fn next_remote_node(
+    mut remote_nodes: Vec<Node>,
+    metrics: &Metrics,
+    tuning_params: KitsuneP2pTuningParams,
+) -> Option<Node> {
+    use rand::prelude::*;
+    let mut rng = thread_rng();
+
+    // Sort the nodes by longest time since we last successfully gossiped with them.
+    // Randomly break ties between nodes we haven't successfully gossiped with.
+    // Note the smaller an Instant the longer it is in the past.
+    remote_nodes.sort_unstable_by(|a, b| {
+        match (
+            metrics.last_success(a.cert()),
+            metrics.last_success(b.cert()),
+        ) {
+            // Choose the smallest (oldest) Instant.
+            (Some(a), Some(b)) => a.cmp(&b),
+            // Put a behind b that hasn't been gossiped with.
+            (Some(_), None) => Ordering::Greater,
+            // Put b behind a that hasn't been gossiped with.
+            (None, Some(_)) => Ordering::Less,
+            // Randomly break ties.
+            (None, None) => {
+                if rng.gen() {
+                    Ordering::Less
+                } else {
+                    Ordering::Greater
+                }
+            }
+        }
+    });
+
+    remote_nodes
+        .into_iter()
+        // Don't initiate with nodes we are currently gossiping with.
+        .filter(|n| !metrics.is_current_round(n.cert()))
```

Comment:
is it possible to do this filter above with the arcset filtering so these nodes don't have to be included in the sorting time?

Reply:
Unless there are going to be a ton of remote nodes, I like it here since it allows this function to take any list of remote nodes without any assumptions, which seems worth the performance hit if it's small.

Reply:
It would require another full iteration to move the filter above which I'm not sure is worth it?